File: library/builtins/traits/trait.rs
use crate::*;

use wipple::*;

pub(crate) fn setup(env: &Env, stack: &Stack) -> Result<()> {
    env.set_variable(stack, "Trait", Value::of(Trait::of::<Trait>()))?;

    // Trait == Text
    env.add_text_relation::<Trait>("trait", stack)?;

    // Trait == Function
    env.add_relation_between(stack, |r#trait: Trait| {
        Function::new(move |value, env, stack| {
            let value = value.evaluate(env, stack)?;

            let trait_value = r#trait.pattern()(&value, env, stack)?
                .ok_or_else(|| error("Cannot use this value to represent this trait", stack))?;

            Ok(Value::new(r#trait.clone(), trait_value.into_owned()))
        })
    })?;

    // Trait == Pattern
    env.add_relation_between(stack, Pattern::for_trait)?;

    Ok(())
}

File: Library/Sources/Traits/Trait.swift
import Wipple

extension Trait: Primitive {}

public extension Value {
    static func checked(trait: Trait, value: Value, _ env: Env, _ stack: Stack) throws -> Self {
        guard let traitValue = try trait.pattern.match(value, env, stack) else {
            throw Exit.error("Cannot use this value to represent this trait", stack)
        }

        return Value(trait: trait, value: traitValue)
    }
}

internal func setupTrait(_ env: Env, _ stack: Stack) throws {
    try env.setVariable("Trait", to: Value(Trait(Trait.self)), stack)

    try env.addRelation(text: "trait", for: Trait.self, stack)

    // Trait == Function
    try env.addRelation(from: Trait.self, to: Function.self, stack) { trait in
        Function { value, env, stack in
            let value = try value.evaluate(env, stack)
            return try .checked(trait: trait, value: value, env, stack)
        }
    }

    // Trait == Pattern
    try env.addRelation(from: Trait.self, to: Pattern.self, stack) { trait in
        Pattern(trait: trait)
    }
}
